### DISCUSSION TITLE
Field specific validation rules to be controlled in the CP

### DISCUSSION THREAD

**Discussion by zizther** [craftcms/cms#9819] [Ideas]
It is often I have to write a custom module to add some validation for some fields to help content editors or users registering make sure they follow the best practices for the site.
A common example is to validate an assets field to only allow one or multiple file extensions (this is more specific than the selection available in the field settings, e.g. a image format maybe selected, but only a jpg extension should be allowed) or test for a maximum file size (the site upload limit maybe higher than a field requires, it would be good to create a limit is this case).
Another example would be username validation where certain rules could be created such as;
- Minimum characters
- Maximum characters
- Contains certain characters
- Can't start with a period

The same thing for passwords where they need to have a minimum length, contain certain characters via a pattern, not contain certain words, etc.

It would show any error messages in the CP as well as the frontend.

I understand it is somewhat easy to create rules and validators in a module, but I wonder how easy it could be to have these as options to add to a field setting and manage via the CP. It may be that basic validation could be set to help cover most scenarios, and extended via a module.
I am also interested to see what others think.

**Comment by brandonkelly** [comment]
What exactly is your suggestion? Not sure I’m following.

**Reply by zizther** [reply]
Sorry for the late reply, I was away.

It would be nice to have a way when managing fields to add additional types of validation, a couple of examples would be;

A assets field, at the moment it is possible to select the file type, if I were to select this as `Image` it would be good to have a way to validating what image extensions could then be uploaded, in this case it may only be for users to upload `.jpg` or `.jpeg` extensions. This would throw a warning when trying to save an image with any other extension. Image weight and dimension could also be validated here, where an image needs to meet a min and max criteria, for example, the image should be no larger than 5mb or should have a minimum width of 1920px and minimum height of 1080px.

A plain text field, at the moment there is some validation which can be implemented, but having the option to add 1 or multiple regex rules within the CP maybe nice, this could be for things such as minimum characters, containing only certain characters, etc.

This could also extend to the user management where it would be nice to set rules for usernames and passwords.
In a couple of sites I have done they've requested that usernames and password follow these rules (which I implemented using a custom module via user rules and validators).

Username rules:
- Min length 5
- Max length 15
- Can only contain alphanumeric, underscore or period characters
- Can't contain reserved words, such as admin or name of the site
- Can't start or finish with a period
- Can't contain 2 or more periods after another

Password rules:
- Min length 6
- Max length 30
- Must contain at least one of each of the following: a number, a lower-case character, an upper-case character, and a special character

For certain validation rules it maybe harder to implement, as some may require something other than a regex to handle the validation, i don't know if there could also be a way round this.

Each rule created could also have an assigned error message, this would show in the CP and frontend if any validation checks fail.

The idea would be to take the need for adding additional rules from custom modules and allow them to be managed per field within the CP.

It maybe some instances above are unique to my use case or not easily possible to implement (with the username and password rules), but I feel having more control for certain fields would save having to write custom rules and checks in a custom module.

**Reply by zizther** [reply]
@brandonkelly Has the explanation made things clearer to understand the requirements of this idea?

**Reply by brandonkelly** [reply]
Sorry, I must have missed that last reply. Makes a lot of sense, thanks!
